pkg/controller/common/rollout/rollout_state.go

```diff
@@ -0,0 +1,197 @@
+package rollout
+
+import (
+	"fmt"
+
+	"k8s.io/klog/v2"
+
+	"github.com/oam-dev/kubevela/apis/standard.oam.dev/v1alpha1"
+)
+
+type rolloutEvent string
+
+const (
+	// rollingSpecVerifiedEvent indicates that we have successfully verified that the rollout spec
+	rollingSpecVerifiedEvent rolloutEvent = "rollingSpecVerifiedEvent"
+
+	// rollingInitializedEvent indicates that we have finished initializing all the workload resources
+	rollingInitializedEvent rolloutEvent = "rollingInitializedEvent"
+
+	// allBatchFinishedEvent indicates that all batches are upgraded
+	allBatchFinishedEvent rolloutEvent = "allBatchFinishedEvent"
+
+	// rollingFailedEvent indicates that the rolling is paused
+	rollingPausedEvent rolloutEvent = "rollingFailedEvent"
+
+	// rollingResumedEvent indicates that the rolling is resumed
+	rollingResumedEvent rolloutEvent = "rollingResumedEvent"
+
+	// rollingFinalizedEvent indicates that we have finalized the rollout which includes but not
+	// limited to the resource garbage collection
+	rollingFinalizedEvent rolloutEvent = "allBatchFinishedEvent"
+
+	// rollingFailedEvent indicates that we encountered an unexpected error during upgrading
+	rollingFailedEvent rolloutEvent = "rollingFailedEvent"
+
+	// initializedOneBatchEvent indicates that we have successfully rolled out one batch
+	initializedOneBatchEvent rolloutEvent = "initializedOneBatchEvent"
+
+	// finishedOneBatchEvent indicates that we have successfully rolled out one batch
+	finishedOneBatchEvent rolloutEvent = "finishedOneBatchEvent"
+
+	// oneBatchAvailableEvent indicates that the batch resource is considered available
+	// this events comes after we have examine the pod readiness check and traffic shifting if needed
+	oneBatchAvailableEvent rolloutEvent = "OneBatchAvailable"
+
+	// batchRolloutWaitingEvent indicates that we are waiting for the approval of resume one batch
+	batchRolloutWaitingEvent rolloutEvent = "batchWaitRolloutEvent"
+
+	// batchRolloutApprovedEvent indicates that we are waiting for the approval of the
+	batchRolloutApprovedEvent rolloutEvent = "batchWaitRolloutEvent"
+
+	// batchRolloutFailedEvent indicates that we are waiting for the approval of the
+	batchRolloutFailedEvent rolloutEvent = "batchRolloutFailedEvent"
+
+	// workloadModifiedEvent indicates that the res
+	workloadModifiedEvent rolloutEvent = "workloadModifiedEvent"
+)
+
+const invalidRollingStateTransition = "the rollout state transition from `%s` state  with `%s` is invalid"
+
+const invalidBatchRollingStateTransition = "the batch rolling state transition from `%s` state  with `%s` is invalid"
+
+// StateTransition is the center place to do rollout state transition
+// it returns an error if the transition is invalid
+// it changes the coming rollout state if it's valid
+// nolint: gocyclo
+func StateTransition(rolloutStatus *v1alpha1.RolloutStatus, event rolloutEvent) error {
```

Comment:
```suggestion
func StateMachine(rolloutStatus *v1alpha1.RolloutStatus, event rolloutEvent) error {
```

What about this name?

Reply:
either way is fine.